v2/tools/generator/internal/astmodel/internal_type_name.go

```diff
@@ -0,0 +1,205 @@
+type InternalTypeName struct {
+	packageReference PackageReference // Note: This has to be a value and not a ptr because this type is used as the key in a map
+	name             string
+}
+
+// MakeInternalTypeName is a factory method for creating a TypeName
+func MakeInternalTypeName(ref PackageReference, name string) InternalTypeName {
+	return InternalTypeName{
+		packageReference: ref,
+		name:             name,
+	}
+}
+
+// Name returns the package-local name of the type
+func (tn InternalTypeName) Name() string {
+	return tn.name
+}
+
+// PackageReference returns the package to which the type belongs
+func (tn InternalTypeName) PackageReference() PackageReference {
+	return tn.packageReference
+}
+
+// WithName returns a new TypeName in the same package but with a different name
+func (tn InternalTypeName) WithName(name string) TypeName {
+	return MakeInternalTypeName(tn.packageReference, name)
+}
+
+// WithPackageReference returns a new TypeName in a different package but with the same name
+func (tn InternalTypeName) WithPackageReference(ref PackageReference) TypeName {
+	return MakeInternalTypeName(ref, tn.name)
+}
+
+// A TypeName can be used as a Type,
+// it is simply a reference to the name.
+var _ Type = InternalTypeName{}
+
+func (tn InternalTypeName) AsDeclarations(codeGenerationContext *CodeGenerationContext, declContext DeclarationContext) []dst.Decl {
+	return AsSimpleDeclarations(codeGenerationContext, declContext, tn)
+}
+
+// AsType implements Type for TypeName
+func (tn InternalTypeName) AsType(codeGenerationContext *CodeGenerationContext) dst.Expr {
+	// If our name is in the current package, we don't need to qualify it
+	if codeGenerationContext.currentPackage.Equals(tn.packageReference) {
+		return dst.NewIdent(tn.name)
+	}
+
+	// Need to ensure we include a selector for that reference
+	packageName, err := codeGenerationContext.GetImportedPackageName(tn.packageReference)
+	if err != nil {
+		panic(fmt.Sprintf(
+			"no reference for %s from %s available in package %s",
+			tn.Name(),
+			tn.packageReference,
+			codeGenerationContext.currentPackage))
+	}
+
+	return astbuilder.Selector(dst.NewIdent(packageName), tn.Name())
+}
+
+// AsZero renders an expression for the "zero" value of the type.
+// The exact thing we need to generate depends on the actual type we reference
+func (tn InternalTypeName) AsZero(definitions TypeDefinitionSet, ctx *CodeGenerationContext) dst.Expr {
+	if IsExternalPackageReference(tn.packageReference) {
+		// TypeName is external, zero value is a qualified empty struct
+		// (we might not actually use this, if the property is optional, but we still need to generate the right thing)
+
+		packageName := ctx.MustGetImportedPackageName(tn.packageReference)
+		return &dst.SelectorExpr{
+			X:   dst.NewIdent(packageName),
+			Sel: dst.NewIdent(fmt.Sprintf("%s{}", tn.Name())),
+		}
+	}
+
+	actualType, err := definitions.FullyResolve(tn)
```

Comment:
Do we want some shared code for bits of this implementation and the `external_type_name`?

Reply:
This looks quite similar to what's over there.